helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java

```diff
@@ -1248,4 +1251,47 @@ private Map<String, Object> getControllerHistory(String clusterId,
   private AclRegister getAclRegister() {
     return (AclRegister) _application.getProperties().get(ContextPropertyKeys.ACL_REGISTER.name());
   }
+
+  /**
+   * Validates the changes to the cluster configuration.
+   *
+   * Specifically checks changes related to topology settings. If topology settings are updated,
+   * ensures all instance configurations align with the new cluster configuration.
+   *
+   * @param clusterName Name of the cluster to validate.
+   * @param configAccessor Accessor to retrieve and update cluster configurations.
+   * @param newClusterConfig The new cluster configuration to validate against.
+   * @param command Type of command triggering this validation (e.g., update, delete).
+   */
+  private void validateClusterConfigChange(String clusterName, ConfigAccessor configAccessor,
+      ClusterConfig newClusterConfig, Command command) {
+    ClusterConfig oldConfig = configAccessor.getClusterConfig(clusterName);
+    ClusterConfig updatedConfig = configAccessor.getClusterConfig(clusterName);
+
+    if (command == Command.delete) {
+      // Since the topology related setting is only in the simple field, we don't need to validate
+      // other fields.
+      for (Map.Entry<String, String> entry : newClusterConfig.getRecord().getSimpleFields()
+          .entrySet()) {
+        updatedConfig.getRecord().getSimpleFields().remove(entry.getKey());
+      }
+    } else {
+      updatedConfig.getRecord().update(newClusterConfig.getRecord());
+    }
+
+    boolean isTopologySettingChanged =
+        (!oldConfig.isTopologyAwareEnabled() && updatedConfig.isTopologyAwareEnabled())
+            || (oldConfig.getTopology() != null && !oldConfig.getTopology().equals(updatedConfig.getTopology()))
+            || (oldConfig.getFaultZoneType() != null && !oldConfig.getFaultZoneType().equals(updatedConfig.getFaultZoneType()));
+
+    if (updatedConfig.isTopologyAwareEnabled() && isTopologySettingChanged) {
+      HelixDataAccessor dataAccessor = getDataAccssor(clusterName);
+      PropertyKey.Builder keyBuilder = dataAccessor.keyBuilder();
+      List<InstanceConfig> instanceConfigs = dataAccessor.getChildValues(keyBuilder.instanceConfigs(), true);
+      for (InstanceConfig instanceConfig : instanceConfigs) {
+        instanceConfig.validateTopologySettingInInstanceConfig(newClusterConfig,
+            instanceConfig.getInstanceName());
```

Comment:
Shouldn't we return some info back to user?
like witch instance config is incompatible etc.

Reply:
Synced offline. The computeInstanceTopologyMapHelper has enough info and it will attached to `ex ` in return info.